gtdynamics/optimizer/ManifoldOptimizerType1.cpp

```diff
@@ -0,0 +1,222 @@
+  constructUnconstrainedValues(ecopt_problem, mopt_problem);
+}
+
+/* ************************************************************************* */
+void ManifoldOptimizerType1::constructManifoldValues(
+    const EqConsOptProblem& ecopt_problem,
+    ManifoldOptProblem& mopt_problem) const {
+  for (size_t i = 0; i < mopt_problem.components_.size(); i++) {
+    // Find the values of variables in the component.
+    const auto& component = mopt_problem.components_.at(i);
+    const Key& component_key = *component->keys_.begin();
+    Values component_values;
+    for (const Key& key : component->keys_) {
+      component_values.insert(key, ecopt_problem.values_.at(key));
+    }
+    // Construct manifold value
+    KeyVector basis_keys;
+    if (p_.cc_params->retract_type ==
+            ConstraintManifold::Params::RetractType::PARTIAL_PROJ ||
+        p_.cc_params->basis_type ==
+            ConstraintManifold::Params::BasisType::SPECIFY_VARIABLES) {
+      basis_keys = (*basis_key_func_)(component);
+    }
+    auto constraint_manifold =
+        ConstraintManifold(component, component_values, p_.cc_params,
+                           p_.retract_init, true, basis_keys);
+    // check if the manifold is fully constrained
+    if (constraint_manifold.dim() > 0) {
+      mopt_problem.values_.insert(component_key, constraint_manifold);
+      mopt_problem.manifold_keys_.insert(component_key);
+    } else {
+      mopt_problem.fixed_manifolds_.insert(component_key, constraint_manifold);
+    }
+  }
+}
+
+/* ************************************************************************* */
+void ManifoldOptimizerType1::constructUnconstrainedValues(
+    const EqConsOptProblem& ecopt_problem,
+    ManifoldOptProblem& mopt_problem) const {
+  // Find out which variables are unconstrained
+  mopt_problem.unconstrained_keys_ = ecopt_problem.costs_.keys();
+  KeySet& unconstrained_keys = mopt_problem.unconstrained_keys_;
+  for (const auto& component : mopt_problem.components_) {
+    for (const Key& key : component->keys_) {
+      if (unconstrained_keys.find(key) != unconstrained_keys.end()) {
+        unconstrained_keys.erase(key);
+      }
+    }
+  }
+  // Copy the values of unconstrained variables
+  for (const Key& key : unconstrained_keys) {
+    mopt_problem.values_.insert(key, ecopt_problem.values_.at(key));
+  }
+}
+
+/* ************************************************************************* */
+void ManifoldOptimizerType1::constructMoptGraph(
+    const EqConsOptProblem& ecopt_problem,
+    ManifoldOptProblem& mopt_problem) const {
+  // Construct base key to component map.
+  std::map<Key, Key> key_component_map;
+  for (const Key& cm_key : mopt_problem.manifold_keys_) {
+    auto cm = mopt_problem.values_.at<ConstraintManifold>(cm_key);
+    for (const Key& base_key : cm.values().keys()) {
+      key_component_map[base_key] = cm_key;
+    }
+  }
+  for (const Key& cm_key : mopt_problem.fixed_manifolds_.keys()) {
```

Comment:
(1) we don't know which constraint manifold each base variable belongs to (that's what the loops are doing: assigning base variables to manifold keys
(2) since a manifold contains several base variables, iterating over manifolds is more efficient.

Reply:
Ah I misread the dependency on `cm`, so nvm. Was doing this review on my phone will at PT.